rclrs/src/node/client.rs

```diff
@@ -0,0 +1,268 @@
+        let mut sequence_number = -1;
+        unsafe {
+            // SAFETY: The request type is guaranteed to match the client type by the type system.
+            rcl_send_request(
+                &*self.handle.lock() as *const _,
+                rmw_message.as_ref() as *const <T::Request as Message>::RmwMsg as *mut _,
+                &mut sequence_number,
+            )
+        }
+        .ok()?;
+        let (tx, rx) = oneshot::channel::<T::Response>();
+        self.futures.lock().insert(sequence_number, tx);
+        Ok(rx.await.unwrap())
+    }
+
+    /// Ask RMW for the data
+    ///
+    /// +----------------------+
+    /// | rclrs::take_response |
+    /// +----------+-----------+
+    ///            |
+    ///            |
+    /// +----------v-----------+
+    /// |   rcl_take_response  |
+    /// +----------+-----------+
+    ///            |
+    ///            |
+    /// +----------v----------+
+    /// |      rmw_take       |
+    /// +---------------------+
+    pub fn take_response(&self) -> Result<(T::Response, rmw_request_id_t), RclrsError> {
+        let mut request_id_out = rmw_request_id_t {
+            writer_guid: [0; 16],
+            sequence_number: 0,
+        };
+        type RmwMsg<T> =
+            <<T as rosidl_runtime_rs::Service>::Response as rosidl_runtime_rs::Message>::RmwMsg;
+        let mut response_out = RmwMsg::<T>::default();
+        let handle = &mut *self.handle.lock();
+        unsafe {
+            // SAFETY: The three pointers are valid/initialized
+            rcl_take_response(
+                handle as *const _,
+                &mut request_id_out,
+                &mut response_out as *mut RmwMsg<T> as *mut _,
+            )
+        }
+        .ok()?;
+        Ok((T::Response::from_rmw_message(response_out), request_id_out))
+    }
+}
+
+impl<T> ClientBase for Client<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    fn handle(&self) -> &ClientHandle {
+        self.handle.borrow()
+    }
+
+    fn execute(&self) -> Result<(), RclrsError> {
+        let (res, req_id) = match self.take_response() {
+            Ok((res, req_id)) => (res, req_id),
+            Err(RclrsError::RclError {
+                code: RclReturnCode::ClientTakeFailed,
+                ..
+            }) => {
+                // Spurious wakeup – this may happen even when a waitset indicated that this
+                // client was ready, so it shouldn't be an error.
+                return Ok(());
+            }
+            Err(e) => return Err(e),
+        };
+        let requests = &mut *self.requests.lock();
+        let futures = &mut *self.futures.lock();
+        if let Some(callback) = requests.remove(&req_id.sequence_number) {
+            callback(&res);
+        }
+        if let Some(future) = futures.remove(&req_id.sequence_number) {
+            future
+                .send(res)
+                .unwrap_or_else(|_| panic!("fail to send response via channel in Client::execute"));
```

Comment:
If it is logically guaranteed to not happen then I'm fine with panicking, but we should have a comment explaining why it won't happen. But here I think it can happen.

The docs say "If the receiving end was dropped before this function was called, however, then `Err(t)` is returned." In this case, the receiving end lives in the `call_async()` function, and the receiving end could be dropped if the user doesn't await that function, or drops a future which awaits that function. In these cases, I think we can assume that the user just isn't interested in the result, and do nothing instead of panicking.

Reply:
Removed panic in e99d1ec